contents/javascript/Javascript-framework-React.md

```diff
@@ -0,0 +1,134 @@
+
+#### React is Component-Based
+
+In the example above, you might have noticed that the application extends a `Component` class. This brings us to the next benefit of React - it is component-based. 
+
+A component is just a JavaScript function that accepts arbitrary input (called `props`) and returns elements describing what should appear on the screen. You can think of a component as an encapsulated piece of your interface that manages its own data and logic. 
+
+Multiple components can be used to compose complex user interfaces. For example, if you have a todo list application, it could have a `Form` component to handle adding new items, and a `List` component to display the items.
+Furthermore, the `List` component could be composed of multiple `ListItem` components, one for each item in the todo list. 
+
+The sandbox below has a simple example to show you how components work:
+
+<box type="info">
+    Notice the HTML-like syntax within the JavaScript method <code>render()</code>. This is <a href="https://reactjs.org/docs/introducing-jsx.html">JSX</a>, and it is a JavaScript syntax extension developed by the React team.
+</box>
+
+<iframe src="https://codesandbox.io/embed/884m88wr90?fontsize=12" style="width:100%; height:400px; border:0; border-radius: 4px; overflow:hidden;" sandbox="allow-modals allow-forms allow-popups allow-scripts allow-same-origin"></iframe>
+
+<br/>
+
+With components written in JavaScript, you can easily pass data through your application and keep state out of the DOM. You can read more about components and props in the React [documentation](https://reactjs.org/docs/components-and-props.html). 
+
+#### React is Fast
+
+Web applications can have a lot of user interaction and data updates. Extensive DOM manipulation can be a performance bottleneck in these applications as the DOM is tree-structured - simple changes at the top can trigger the entire application to get updated. 
+
+React solves this problem by using a virtual DOM. The virtual DOM is a JavaScript object that is kept in the memory of your application. In other words, it is a virtual representation of the actual DOM.
+
+Any updates to your application will first be made to the virtual DOM. Then, React will compare the virtual DOM with the actual DOM using a [diffing algorithm](https://reactjs.org/docs/reconciliation.html) to find out what changes have to be made to the actual DOM. 
+Finally, React figures out how to update the actual DOM in the most effective way possible to guarantee minimum update time. This results in better performance and a cleaner user experience, even with data-intensive applications.
+
+Those are the three main benefits of React. Other advantages are as follows:
+ - React only allows data to flow downwards (one-way data binding), which makes your application easier to debug.
+ - It is a constantly developing open-source library that is backed by Facebook, so you can be sure of getting community support.
```

Comment:
I think you can add something like `Since it is a constantly developing open-source library that is backed by Facebook, it is constantly being updated with new features/performance optimisations?`.
Otherwise, the first statement doesn't add any value. In fact, it is kinda mentioned as a disadvantage below

Reply:
You're right, I'll update the statement to `constantly being updated with new features/performance optimisations`.